feat(blog): 6.4 Article (#2423)
withastro/astro.build: src/content/blog/astro-640.mdx

---
title: 'Astro 6.4'

> description: 'Astro 6.4 introduces a new pluggable Markdown processor API, a Rust-based Markdown processor for faster builds, and first-party CDN cache providers for your favorite platforms.'

Reviewer: Just a reminder we'll need to update that. Feel free to use something else!
```suggestion
description: 'Astro 6.4 introduces a new pluggable Markdown processor API, a Rust-based Markdown processor for faster builds, and helpers to wire up experimental advanced routing with Cloudflare.'
```

homepageLink:
  title: 'Astro 6.4'
  subtitle: 'Available now!'
publishDate: '2026-05-28'
authors:
  - erika
coverImage: '/src/content/blog/_images/astro-640/blog-post-astro-6.4.webp'
socialImage: '/src/content/blog/_images/astro-640/og-astro-6.4.jpg'
lang: 'en'
related:
  - astro-630
  - astro-620
---

import Mention from '/src/components/Mention.astro';

> **Astro 6.4 is here!** On today's menu: a pluggable Markdown pipeline, a new Rust-based Markdown processor for you to try, and first-party CDN cache providers for Cloudflare, Netlify, and Vercel.

Reviewer: Same here, feel free to use something else that better suits your tone:
```suggestion
**Astro 6.4 is here!** On today's menu: a pluggable Markdown pipeline, a new Rust-based Markdown processor for you to try, and helpers to facilitate the use of experimental advanced routing with Cloudflare.
```

---
title: 'Astro 6.4'
description: 'Astro 6.4 introduces a new pluggable Markdown processor API, a Rust-based Markdown processor for faster builds, and first-party CDN cache providers for your favorite platforms.'
homepageLink:
  title: 'Astro 6.4'
  subtitle: 'Available now!'
publishDate: '2026-05-28'
authors:
  - erika
coverImage: '/src/content/blog/_images/astro-640/blog-post-astro-6.4.webp'
socialImage: '/src/content/blog/_images/astro-640/og-astro-6.4.jpg'
lang: 'en'
related:
  - astro-630
  - astro-620
---

import Mention from '/src/components/Mention.astro';

**Astro 6.4 is here!** On today's menu: a pluggable Markdown pipeline, a new Rust-based Markdown processor for you to try, and first-party CDN cache providers for Cloudflare, Netlify, and Vercel.

Explore what's new in this release:

- [**New `markdown.processor` API**](#new-markdownprocessor-api)
- [**Faster Markdown builds with Sätteri**](#faster-markdown-builds-with-sätteri)

> - [**CDN cache providers for Cloudflare, Netlify, and Vercel**](#cdn-cache-providers-for-cloudflare-netlify-and-vercel)

Author: ```suggestion
```

const data = await fetchProducts();
---
<ProductList items={data} />
```

The equivalent providers are available as `cacheNetlify()` from `@astrojs/netlify/cache` and `cacheVercel()` from `@astrojs/vercel/cache`. All three set the appropriate platform-specific cache headers (`Cloudflare-CDN-Cache-Control`, `Netlify-CDN-Cache-Control`, `Vercel-CDN-Cache-Control`, and so on) and support both tag-based and path-based invalidation:

```ts title="src/pages/api/purge.ts"
export async function POST({ request, cache }) {
  await cache.invalidate({ tags: ['products'] });
  return new Response('Purged');
}
```

You can also set cache rules for groups of routes from your config using `routeRules`:

```js title="astro.config.mjs"
experimental: {
  cache: { provider: cacheCloudflare() },
  routeRules: {
    '/products/*': { maxAge: 3600, tags: ['products'] },
    '/api/*': { maxAge: 60, swr: 600 },
  },
},
```

Cache hits are served directly from the CDN's edge, so your function or worker is never invoked. For more information on enabling and using this feature, see the [experimental route caching docs](https://docs.astro.build/en/reference/experimental-flags/route-caching/).

> 

Reviewer 2: ```suggestion
```

---
title: 'Astro 6.4'
description: 'Astro 6.4 introduces a new pluggable Markdown processor API, a Rust-based Markdown processor for faster builds, first-party CDN cache providers for your favorite platforms, and a stable custom logger.'
homepageLink:
  title: 'Astro 6.4'
  subtitle: 'Available now!'
publishDate: '2026-05-28'
authors:
  - ema
  - erika
  - matt
coverImage: '/src/content/blog/_images/astro-640/blog-post-astro-6.4.webp'
socialImage: '/src/content/blog/_images/astro-640/og-astro-6.4.jpg'
lang: 'en'
related:
  - astro-630
  - astro-620
---

import Mention from '/src/components/Mention.astro';

> **Astro 6.4 is here!** On today's menu: Astro's Markdown pipeline is now pluggable. A new `markdown.processor` config option lets you swap out the default unified pipeline, a new Rust-based processor for you to try out, first-party CDN cache providers for Cloudflare, Netlify, and Vercel, and stabilized custom logging.

Reviewer 3: That sentence reads a little weird as the `lets you` sets up another type of follow-up. I think adding a `there's` will improve readability. But this suggestion is rather opinionated.

```suggestion
**Astro 6.4 is here!** On today's menu: Astro's Markdown pipeline is now pluggable. A new `markdown.processor` config option lets you swap out the default unified pipeline, there's a new Rust-based processor for you to try out, first-party CDN cache providers for Cloudflare, Netlify, and Vercel, and stabilized custom logging.
```

Reviewer 3: This is resolved, but I can not mark it as resolved for some reason...

Explore what's new in this release:

- [**New `markdown.processor` API**](#new-markdownprocessor-api)
- [**Faster Markdown builds with Sätteri**](#faster-markdown-builds-with-sätteri)
- [**CDN cache providers for Cloudflare, Netlify, and Vercel**](#cdn-cache-providers-for-cloudflare-netlify-and-vercel)
- [**Stable custom logger**](#stable-custom-logger)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## New `markdown.processor` API

Astro's Markdown pipeline has always been built around [unified](https://unifiedjs.com/) (remark, rehype, and friends). That ecosystem is powerful, having thousands of plugins available, but perhaps it's not up to your taste. Astro 6.4 adds a new `markdown.processor` configuration option that lets you swap that pipeline out entirely.

> The default processor is still `unified()`, so existing projects keep working without any changes. Your remark and rehype plugins continue to behave the same. They just live on the processor now, instead of the top-level `markdown` config:

Reviewer 3: A little personal rephrase — feel free to decide if you like it more or not:

```suggestion
The default processor is still `unified()`, so existing projects keep working without any changes. Your remark and rehype plugins continue to behave the same — they are just configured directly on the processor now, instead of in the top-level `markdown` config:
```

```js title="astro.config.mjs"
import { defineConfig } from 'astro/config';
import { unified } from '@astrojs/markdown-remark';
import remarkToc from 'remark-toc';

export default defineConfig({
  markdown: {
    processor: unified({
      remarkPlugins: [remarkToc],
    }),
  },
});
```

> The existing top-level `markdown.remarkPlugins`, `markdown.rehypePlugins`, `markdown.remarkRehype`, `markdown.gfm`, and `markdown.smartypants` options still work, but they're now deprecated in favor of passing them to `unified({...})`, and will be removed in Astro 8.0.

Reviewer 3: Did I just read Astro 8.0? 👀

Author: There's nothing planned for it, it's just the major after the next one. Had this shipped earlier presumably we would've removed in 7 instead

For more information, see the updated [Markdown guide](https://docs.astro.build/en/guides/markdown-content/).

## Faster Markdown builds with Sätteri

A pluggable processor API also means Astro can ship other Markdown processors. Astro 6.4 introduces a new `@astrojs/markdown-satteri` package, with a new processor based on [Sätteri](https://satteri.bruits.org/), a Markdown and MDX pipeline written in Rust.

> Sätteri is much faster than the default unified-based pipeline, and it implements many Markdown features natively that previously required plugins. To try it out, install the package and set it as your processor:

Reviewer 3: Since `"much faster"` is mentioned. 
As a reader, I would love to see some stats here of how fast Sätteri actually is compared to the unified pipeline and what impact it has on Astro's markdown pipeline.

(If such a graphic is available / possible in time)

const data = await fetchProducts();
---
<ProductList items={data} />
```

The equivalent providers are available as `cacheNetlify()` from `@astrojs/netlify/cache` and `cacheVercel()` from `@astrojs/vercel/cache`. All three set the appropriate platform-specific cache headers (`Cloudflare-CDN-Cache-Control`, `Netlify-CDN-Cache-Control`, `Vercel-CDN-Cache-Control`, and so on) and support both tag-based and path-based invalidation:

```ts title="src/pages/api/purge.ts"
export async function POST({ request, cache }) {
  await cache.invalidate({ tags: ['products'] });
  return new Response('Purged');
}
```

You can also set cache rules for groups of routes from your config using `routeRules`:

```js title="astro.config.mjs"
experimental: {
  cache: { provider: cacheCloudflare() },
  routeRules: {
    '/products/*': { maxAge: 3600, tags: ['products'] },
    '/api/*': { maxAge: 60, swr: 600 },
  },
},
```

Cache hits are served directly from the CDN's edge, so your function or worker is never invoked. For more information on enabling and using this feature, see the [experimental route caching docs](https://docs.astro.build/en/reference/experimental-flags/route-caching/).

> ## Stable custom logger

Reviewer 3: Not sure if this section will stay in here, as the logger docs PR recently got added to the `7.0` milestone: https://github.com/withastro/docs/pull/13907

If you were already using the experimental flag, just drop the `experimental.` prefix:

```diff lang="js" title="astro.config.mjs"
import { defineConfig } from 'astro/config';

export default defineConfig({
-  experimental: {
-    logger: {
-      entrypoint: '@org/custom-logger',
-    },
-  },
+  logger: {
+    entrypoint: '@org/custom-logger',
+  },
});
```

For more details, see the [logger reference](https://docs.astro.build/en/reference/configuration-reference/#logger).

## Other improvements

For a complete list of bug fixes and smaller improvements, see [the full changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md).

## Community

The Astro core team is:

<Mention name="alex" />, <Mention name="armand" />, <Mention name="chris" />,
<Mention name="ema" />,
<Mention name="erika" />, <Mention name="florian" />, <Mention name="fred" />,
<Mention name="hideoo" />,
<Mention name="luiz" />, <Mention name="matt" />, <Mention name="matthew" />,
<Mention name="reuben" />,
<Mention name="sarah" />, and <Mention name="yan" />.

> Thanks to all the other contributors who helped make Astro 6.4 possible with code and docs additions and improvements.

Reviewer 3: Not sure if you wanna continue mentioning everybody, but I copied the list from the Astro 6.3 blog.

```suggestion
Thanks to all the other contributors who helped make Astro 6.4 possible with code and docs additions and improvements.

[0x K.](https://github.com/0xbejaxer), [AceCodePt](https://github.com/AceCodePt), [Adam Chalemian](https://github.com/adamchal), [AitorMT](https://github.com/AitorMT), [atsbob](https://github.com/atsbob), [Calvin Liang](https://github.com/Calvin-LL), [Chan](https://github.com/enjoyandlove), [done](https://github.com/dominikg), [Eveeifyeve](https://github.com/Eveeifyeve), [Felix Schneider](https://github.com/trueberryless), [Felmon](https://github.com/felmonon), [fkatsuhiro](https://github.com/fkatsuhiro), [junetpoint](https://github.com/junetpoint), [Junseong Park](https://github.com/jsparkdev), [knj](https://github.com/jp-knj), [ld-web](https://github.com/ld-web), [Lee Freeman](https://github.com/leefreemanxyz), [Louis Escher](https://github.com/louisescher), [Luca Faccio](https://github.com/Sunny-Pirate), [Maxim Slobodchikov](https://github.com/maximslo), [oandy-rgb](https://github.com/oandy-rgb), [ocavue](https://github.com/ocavue), [Ossaid](https://github.com/ossaidqadri), [Patrick Linnane](https://github.com/p-linnane), [Quetzal Rivera](https://github.com/Eptagone), [Rafael Yasuhide Sudo](https://github.com/rururux), [randomguy-2650](https://github.com/randomguy-2650), [Rayan Salhab](https://github.com/cyphercodes), [Rodrigo Santos](https://github.com/rodrigosdev), [Roman](https://github.com/gameroman), [Sigma](https://github.com/sigma7863), [Utpal Sen](https://github.com/senutpal), [vrabe](https://github.com/vrabe), and [web-dev0521](https://github.com/web-dev0521)
```

Or do you always generate that list per minor? 🤔 
In that case, just ignore this comment 👍

Author: We generate those automatically!
